### DISCUSSION TITLE
Update internal pip version

### DISCUSSION THREAD

**Discussion by sswatson** [astral-sh/rye#1300] [Q&A]
The pip used internally by rye is currently pinned to 23.3.2 (in the file piptools.rs), and that's a problem for me because the dependency resolution logic in that version is apparently faulty in a way that breaks projects that include the main library I'm working on. Specifically, it gets NumPy 2 even though some of the deps haven't been upgraded yet.

Is there a solution to this problem short of editing that hard-coded version number and re-compiling rye?

**Comment by bluss** [comment]
That sounds like a problem that is worthy of reporting an issue.

But I also think that the pip side of rye has not gotten any attention since Rye started using uv as the backend, so uv should normally work better. Is there a reason why you have configured rye to keep using pip-tools? If there's a reason that would be good to mention too (because removing the pip-tools support is also an alterantive that's being looked at in open PRs.)

There are always hacky solutions - they come with disclaimers that they don't work very well. The self venv is located at ~/.rye/self and it can be inspected and changed. Rye will recreate it under certain circumstances. Also look at ~/.rye/pip-tools which is the pip tools specific environment.

**Reply by sswatson** [reply]
Thanks; that's really helpful. I didn't realize that I was using a non-default configuration, and I'd prefer to be using uv actually. I've been using rye since well before the uv days, so the reason I'm not already on it is just that I didn't ever go back and edit `~/.rye/config.toml`.

That said, I tried adding `use-uv = true`, and while it's delightfully fast, it has the same problem that the older pip version does. Namely, if you `rye add relationalai`, you get `pyarrow==14.0.2` along with `numpy==2.0.1` and those two things apparently aren't compatible. Doing `rye add 'numpy<2'` fixes the problem. (Note: the `relationalai` package itself doesn't matter here; it's just the only way I know of to trigger that combination of versions.)

Feels like the uv repo would be the right one to file that issue on?

**Reply by charliermarsh** [reply]
Can you help me come up with a full reproduction for this? `rye add relationalai` gives me `numpy==1.26.3` and `pyarrow==15.0.2`.

**Reply by sswatson** [reply]
OK, that's even more interesting than the original problem. I would have thought the set of resolved versions would be a deterministic function of `rye --version` and the `use-uv` setting in `config.toml`. (Well, it could also depend on time, but I just confirmed that I can reproduce the original behavior, so that doesn't seem like what's going on.)

I opened an issue [here](https://github.com/astral-sh/rye/issues/1312) with as many repro details as I could think of.

**Comment by charliermarsh** [comment]
Doesn't solve your problem _right now_, but I'd be happy to update the internal version in the next release -- not a problem.

**Reply by sswatson** [reply]
I'm not in a particular hurry, so this would be a great solution. Thanks!

**Reply by charliermarsh** [reply]
https://github.com/astral-sh/rye/pull/1304
